akka-projection-cassandra/src/main/scala/akka/projection/cassandra/internal/CassandraProjectionImpl.scala

```diff
@@ -146,6 +176,12 @@ import akka.stream.scaladsl.Source
           }
       case AtMostOnce =>
         // prevent usage of retries for at-most-once
+        val handler = handlerStrategy match {
+          case SingleHandlerStrategy(handler) => handler
+          case _                              =>
+            // not possible
+            throw new IllegalStateException("Unsupported combination of atMostOnce and grouped")
```

Comment:
This makes me wondering if we are not cutting the abstracting on the wrong place. 
Maybe we could move that whole logic down to the handler itself. 

Each implementation has its own handlers implementations, then, for Cassandra, we could have three kinds of Handlers (AtLeastOnceSingle, AtLeastOnceGrouped and AtMostOnce). The AtLeastOnce variants would also implement `HandlerRecovery`.

For the user, it will have the impact that they need to decide earlier, but not a big issue. It should be easy to move from AtLeastOnceSingle to AtMostOnce. The user logic doesn't change.

Reply:
but note that is is only internals, it's not possible for the user to create this combination of atMostOnce and grouped

Reply:
Hmm, this has some consequences for things like ActorHandler. 

To take the path I suggested, we will need to make a distinction from `Handler` and `HandlerExecutionStrategy`. 

`Handler` is the user function and can be a pure function or the ActorHandler variant (or others), the `HandlerExecutionStrategy` is dependent on the underlying tech and options it gives, ie: AtLeastOnce (grouped or not), AtLostOnce, ExactlyOnce. 

Maybe `HandlerRecovery` should not be part of `Handler`, but of `HandlerExecutionStrategy`.

Reply:
> but note that is is only internals, it's not possible for the user to create this combination of atMostOnce and grouped

I see. Still, I think the point about `HandlerRecovery` is still valid. I think it's a property of `AtLeastOnce` and `ExactlyOnce` and not from `Handler`. 

As of today, I can define a handler with recovery strategy and run it with atMostOnce.

Reply:
we don't need to expose it as a strategy, but maybe as specific settings.

Reply:
When I worked on the HandlerRecovery I was thinking that it was maybe overkill to be able to decide based on the failed envelope inside the handler. Maybe it would be enough with a setting for the projection and not have it as a function.

Just
```
CassandraProjection.atLeastOnce(...).withRecovery(HandlerRecoveryStrategy.retryAndFail(...)
```

Then we can make that only possible for atLeastOnce and not atMostOnce.

Similar to what we talked about for the atLeastOnce parameters.

Reply:
this was captured in issue #128